File: Graph/Java/As Far from Land as Possible.java
// Problem link: https://leetcode.com/problems/as-far-from-land-as-possible/description/
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int maxDistance(int[][] grid) {
        int rows = grid.length, cols = grid[0].length, result = -1, rw, cl, size;
        int[][] visited = new int[rows][cols];
        Queue<int[]> queue = new LinkedList<>();
        int[][] directions = new int[][] {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 1) {
                    queue.offer(new int[] {row, col, 0});
                    visited[row][col] = 1;
                }
            }
        }

        while(!queue.isEmpty()) {
            size = queue.size();

            for(int itr = 0; itr < size; itr++) {
                int[] store = queue.poll();

                for(int[] dir : directions) {
                    rw = store[0] + dir[0];
                    cl = store[1] + dir[1];

                    if(rw >= 0 && cl >= 0 && rw < rows && cl < cols && visited[rw][cl] == 0) {
                        queue.offer(new int[] {rw, cl, store[2] + 1});
                        result = Math.max(result, store[2] + 1);
                        visited[rw][cl] = 1;
                    }
                }
            }
        }

        return result;
    }
}

File: Graph/Python/As Far from Land as Possible.py
# Problem link: https://leetcode.com/problems/as-far-from-land-as-possible/description/
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        rows, cols, result, queue = len(grid), len(grid[0]), -1, deque()
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        visited = [[0] * cols for _ in range(rows)]


        for row in range(rows):
            for col in range(cols):
                if grid[row][col]:
                    queue.append((row, col, 0))
                    visited[row][col] = 1
        
        while queue:

            for _ in range(len(queue)):
                row, col, dist = queue.popleft()

                for dr, dc in directions:
                    rw, cl = row + dr, col + dc

                    if rw >= 0 and cl >= 0 and rw < rows and cl < cols and not visited[rw][cl]:
                        queue.append((rw, cl, dist + 1))
                        result = max(result, dist + 1)
                        visited[rw][cl] = 1
            
        return result
